src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java

```diff
@@ -319,191 +316,91 @@ static void verifyColumnMasterKeyMetadata(SQLServerConnection connection, SQLSer
 
-                    return token;
-                }
-            } catch (Exception e) {
-                retry++;
-                // Below code applicable only when !isAzureFunctcion (VM)
-                if (retry > maxRetry) {
-                    // Do not retry if maxRetry limit has been reached.
-                    break;
-                } else {
-                    try {
-                        int responseCode = connection.getResponseCode();
-                        // Check Error Response Code from Connection
-                        if (GONE == responseCode || TOO_MANY_RESQUESTS == responseCode || NOT_FOUND == responseCode
-                                || (INTERNAL_SERVER_ERROR <= responseCode
-                                        && NETWORK_CONNECT_TIMEOUT_ERROR >= responseCode)) {
-                            try {
-                                int retryTimeoutInMs = retrySlots.get(ThreadLocalRandom.current().nextInt(retry - 1));
-                                // Error code 410 indicates IMDS upgrade is in progress, which can take up to 70s
-                                retryTimeoutInMs = (responseCode == 410
-                                        && retryTimeoutInMs < imdsUpgradeTimeInMs) ? imdsUpgradeTimeInMs
-                                                                                   : retryTimeoutInMs;
-                                Thread.sleep(retryTimeoutInMs);
-                            } catch (InterruptedException ex) {
-                                // re-interrupt thread
-                                Thread.currentThread().interrupt();
-
-                                // Throw runtime exception as driver must not be interrupted here
-                                throw new RuntimeException(ex);
-                            }
-                        } else {
-                            if (null != msiClientId && !msiClientId.isEmpty()) {
-                                throw new SQLServerException(
-                                        SQLServerException.getErrString("R_MSITokenFailureImdsClientId"), null);
-                            } else {
-                                throw new SQLServerException(SQLServerException.getErrString("R_MSITokenFailureImds"),
-                                        null);
-                            }
-                        }
-                    } catch (IOException io) {
-                        // Throw error as unexpected if response code not available
-                        throw new SQLServerException(SQLServerException.getErrString("R_MSITokenFailureUnexpected"),
-                                null);
-                    }
-                }
-            } finally {
-                if (connection != null) {
-                    connection.disconnect();
-                }
+        if (!accessTokenOptional.isPresent()) {
+            if (connectionlogger.isLoggable(java.util.logging.Level.FINE)) {

```

Comment:
Similar to the ManagedIdentity method, if we get no result and no error, we need to throw our own exception. We should never return a null result when getFedAuthToken is called.